### PR TITLE
Add support for decoding jpeg's with arithmetic coding

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This PR adds support for decoding jpeg's with arithmetic coding.

### CODE REVIEW

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticStatistics.cs
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticStatistics.cs:8]: Can this be written as
```c#
internal unsafe struct ArithmeticStatistics
{
    private fixed byte statistics[256];

    public ArithmeticStatistics(bool dc, int identifier)
    {
        this.IsDcStatistics = dc;
        this.Identifier = identifier;
    }

    public bool IsDcStatistics { get; private set; }

    public int Identifier { get; private set; }

    public ref byte GetReference() => ref this.statistics[0];

    public ref byte GetReference(int offset) => ref this.statistics[(uint)offset];

    public void Reset()
    {
        Unsafe.InitBlockUnaligned(ref this.GetReference(), 0x00, (uint)(this.IsDcStatistics ? 64 : 256));
    }
}
```
?
* avoids the additional allocation for each stats-object
* downside: size is _fixed_ to 256, so potentially the list will be larger
* [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNXjxldskht1zYAZjAa4MUDmAwMAglF4YAFgL5hZGbH0vLcNADeNAyhDGJOCj7mJrwIMAAmDMAAnhjmlj68fmC4ANqkAKxIALoa1CFhxADMDk6u7speWTm4ABTAEBAGCWBoDNp2vAkwPLyxMgCUlaHB1GELDK7ZLACSuAAint6+HrgMALwMveWLYcu4ayNjE1CHA9d8t6dhAL6a81W1nd0M61vNXb+BiBBgAcxgGDU4Ui0QskOh7wqn1CNQGPD+j3GvBkIPBCJhvCi6XhUIYSJmTFqsBMyTS5gA4pC2DAzLAuCo2pNDgA+Bg0pYuFaZIG5PJ0MofBZogWpElMjAstmjTmDBgQEwmXCQ7kHPkCi4sEXZPZ5NocQaTDVayGS5HS2rEFAMFnajBcylzM6hACqXGMZjWXGcACFJBAwABrP36XhgriJNoGoWXBVKmQqmBc/p0BB0Oj9c2WtqG/7bFp7BgAfgYqAYIAYRSQk0mL1CSNeQA) is better than [with the class](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKAEsA7DGKe7AGwAIyOBBKWjAAt8MDLTABlDNjG4xYXDQ7KOAbyUrNAB34A3GTA6xsAEwj02ATw7BLTANoBdDnJm0543AG4aGzcuIAZl5+IRF5KTcPBQAKYAgIThMwNA4GDDSTGEZaADNaZgBKP391an8KjiF3ADoASVwAEUlpWXlcDgBeDmSfcsrNatx6rJz85i7M7LFxqD6BwcFa1zbPSeSOAH4OehgAdxs7GHtUZxAd/cOHUgBWJEd5yoBfX36KoJsEzgbmyNWFNQcADmoi8HB0tH0TBcoI4L2oJU0H3SHDqoxmBSggJBGDBEKhhlwsPhiJUH1guSuhgA4qIAEowXLMbJgGAxQpdAB8RkZVSWwxW7na9gADA9XgNybzbNDaRgGUzYPRWTEURBcrkiRgOZ1uRS+ctWkLPPZ1ZrROKEW9/B9iCgOAyteyuQaBUbosMeLgJFpsPR2TUAMJsGDYKDsx7KJ5AA=) especially for `GetReference` and `Reset` (which are the hottest methods here?)

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/IJpegComponent.cs
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/IJpegComponent.cs:16]: The existing interface members don't have the access modifier specified.
It's not strictly needed, but C# allows this (since C# 8 I think).

At least it should be consistent withing this file.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:1109]: To double check: stream (in general) isn't guaranteed to read exactly the requested length, it may read up to length bytes. So this can be a latent bug, and usually there's a `while (bytesRead < length)`-loop.

I see the check below, and assume that BufferedReadStream's implemented takes this into mind (via CopyTo over there?).

So this is more or less a question to avoid a latent, nasty to find bug.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:1131]:
```suggestion
            if ((frameHeight | frameWidth) == 0)
```
produces [less code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBgDEIEAELYoNAN40GtpruDHJh4+QAUvHgwBmUbAIAJGF4AcwALDDQGTwwfPwEAdV4AEwwwgEoGAF4APjj/GCDQiOyshkZCQnzElLTS8o1qO3sGRwhnIwhSDy9fAqLwyOje+Jgk1IzsvLc+wODBhirZsdrJrLK6RoBfIA) and avoids a branch.

Note: current JIT will do this optimizaiton for use, but older runtimes miss that optimization.

**Review by gfoidl — COMMENTED** [review]
I always enjoy reading such fine code.
At the same time I'm sorry for so much comments.

**Comment by br3aker** [comment]
I know that I'm asking for too much but can we postpone this PR from merging for a bit?
There's so much virtual types existing in current jpeg decoder and new ones from this PR which already led to explicit casting. I'm working on the downscaling decoding for jpeg which would also need some sort of a virtuality for both the huffman decoder and component post processing which combined can cause a mess IMO.

I'll push some fixes on your branch if you do't mind.

**Comment by brianpopow** [comment]
@gfoidl thanks for your review, its always appreciated!

@br3aker:

> I know that I'm asking for too much but can we postpone this PR from merging for a bit?

Its not a problem, there is no need to rush this PR.

> I'll push some fixes on your branch if you don't mind.

I would not mind, if you push directly to this branch, your help is very welcome, but I dont think you have permission to do so.
I am pretty sure its somehow possible to give you those write access to this branch, but I did not see how at the moment.
You will have to create a regular PR with this branch as the target branch instead of `main`.

**Comment by br3aker** [comment]
Code itself is very good, it's just those tiny little spots I'd want to fix but they do not block merging so it's okay. I will create a PR with general refactoring of the decoder part.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:1109]: Yeah `BufferedReadStream` uses a while loop to read internally.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs
Comment (brianpopow) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs:472]: @gfoidl I dont understand the suggestion: How can I use `uint` here, when `Unsafe.Add` expects `int`?

Comment (gfoidl) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs:472]: Here it's for the overload `Unsafe.Add<T>(ref T source, nuint elementOffset)` and uses the implicit conversion from `uint` -> `nuint`.
So the long form of the suggestion is
 ```suggestion
                                    ref Unsafe.Add(ref blockRef, (nuint)(uint)blockCol),
```
but as the compiler does the second conversion for us, I've shortened it to
```suggestion
                                    ref Unsafe.Add(ref blockRef, (uint)blockCol),
```

---

In essence, and to avoid the `movsxd` (the sign extending move), we need to convert the `int` to any of `IntPtr` / `nint` or `UIntPtr` / `nuint`. As first step here is to prove to the JIT that there's only positive numbers by the `uint`-cast, which elides the `movsxd` emitting. 
The actual target type (`nint` or `nuint` respectively there IntPtr base variants) doesn't really matter then.

Comment (brianpopow) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs:472]: Thank you for the explanation. Now I see why it does not work: The uint overload is only available on .net6.0, so cant use that here.

Comment (gfoidl) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs:473]: Ah, sorry, didn't check that. The use `(nint)(uint)` for the cast.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs
Comment (brianpopow) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs]: For some reason this change introduced a access violation issue. I cant spot why that is:

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
```

See CI output for full stack trace.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (br3aker) [inline src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:111]: I don't like this arithmetic coding specific property is leaked in the 'master' class, same goes for reset interval which should only be used inside `scanDecoder` but it's a broad question for some further refactoring - I'll work on it right after scaled decoding PR.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/JpegFrame.cs
Comment (br3aker) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/JpegFrame.cs:89]: As far as I remember, `IJpegComponent` interface is used in tests for libjpeg output comparison but I think we shouldn't use this interface everywhere - [this leads to virtual calls for every component field getter ](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmAjAWAChMBmAAk1wguzIGEyBvIs1i8gSwDs4BTAJwBmiAMa8yASToB7VAAdpXXjzIs2zQmy1lg06QBsyAMX2IA5kzO84AbgC+a1g8KP2NerIVKVISTPmKynCuGtpspDp6hibmltb2rs5arhGU1LoGZADi1jFmABRSnoEqIp4AlGQAvAB8ZGXyAHR5NinkaZGZOXB5+f5eQfUV1XUNcs2mZq2EdkA==).

**Review by br3aker — CHANGES_REQUESTED** [review]
Looks good overall but please revert component type to `JpegComponent` in `JpegDecoderCore`.

**Review by br3aker — APPROVED** [review]

**Comment by br3aker** [comment]
@brianpopow hi! Anything blocking this PR from being merged?

**Comment by brianpopow** [comment]
> @brianpopow hi! Anything blocking this PR from being merged?

@br3aker: I think this can be merged now. 
@JimBobSquarePants do you have any objection on merging this?

**Comment by JimBobSquarePants** [comment]
Nope. Rock on

**Pull request merged**
